CRM/Civicase/Service/CaseMenu/Route.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+use CRM_Civicase_Service_CaseCategoryPermission as CaseCategoryPermission;
+
+/**
+ * Base class for Menu Routes.
+ *
+ * @method string getLabel()
+ * @method string getName()
+ * @method string getUrl()
+ * @method string getPermission()
+ * @method string getPermissionOperator()
+ * @method string getHasSeparator()
+ */
+abstract class CRM_Civicase_Service_CaseMenu_Route {
+
+  /**
+   * Case Type Category Name.
+   *
+   * @var string
+   */
+  protected $categoryName;
+
+  /**
+   * Case Type Category ID.
+   *
+   * @var int
+   */
+  protected $categoryId;
+
+  /**
+   * Array of permissions for the associated Category.
+   *
+   * @var array
+   */
+  protected $permissions;
+
+  /**
+   * Properties with default value overwritten.
+   *
+   * All the attributes of the route class can be overwritten here, filling
+   * this array using the appropriate setter.
+   *
+   * @var array
+   */
+  protected $overwrittenProperties;
+
+  /**
+   * Constructor of the route.
+   *
+   * @param string $categoryName
+   *   Case Type Category Name for the route.
+   * @param int $categoryId
+   *   Case Type Category Id for the route.
+   */
+  public function __construct(string $categoryName, int $categoryId) {
+    $this->permissions = (new CaseCategoryPermission())->get($categoryName);
+
+    $this->categoryName = $categoryName;
+    $this->categoryId = $categoryId;
+
+    $this->overwrittenProperties = [];
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function setOverwrittenProperties($overwrittenProperties) {
+    $this->overwrittenProperties = $overwrittenProperties;
+
+    return $this;
+  }
+
+  /**
+   * Magic method for handling the overwritten properties.
+   *
+   * If a properties was overwritten, that value will be returned. If not, the
+   * underlying method will be called.
+   */
+  public function __call($method, $arguments) {
+    $property = strtolower(str_replace('get', '', $method));
+
+    if (isset($this->overwrittenProperties[$property])) {
+      return $this->overwrittenProperties[$property];
+    }
+
+    if (method_exists($this, $property)) {
+      return $this->$property();
+    }
+  }
+
+  /**
+   * Return array with all route data.
```

Comment:
Should not we document return value?

Reply:
Yes. In some cases can be omited, if it is not unique, but in the cases you pointed is ok. I added this and other more here: https://github.com/compucorp/uk.co.compucorp.civicase/pull/756/commits/66424d5a429d06bf2c4ffbd52a730bd445cc825d (I will mark the other comments as resolved too)